### PR TITLE
Add Benchmarking CI

### PR DESCRIPTION
**Pull request by christiangnrd**
Using a JuliaGPU branch instead of from my fork. See #419 for the start of this branch.

**Original text:**
Copying over the benchmarks from CUDA.jl.

I'm not sure if I converted them properly. Some function seem to be missing Metal implementations (like `reverse`).

The final (and biggest) problem is how inconsistent these results have been. Simply rerunning the benchmarks on the same code gives some huge performance differences. There is always at least one benchmark that is >20% slower or faster. The benchmarks seem to be way more consistent on the runners. Only a few seem to have big variance.

See https://github.com/JuliaGPU/Metal.jl/pull/418#issuecomment-2355262678

Todo:
- [x] Implement actual CI benchmark solution from GemmKernels.jl
- [ ] Ensure Benchmarks are correct

### CODE REVIEW

**Comment by christiangnrd** [comment]
The culprit was whitespace... This is (one of the many reasons) why julia > python.

**Comment by christiangnrd** [comment]
@maleadt Do you know how I can set up the buildkite token on the Github Actions side?

**Comment by maleadt** [comment]
I can add a token. Which permissions do you need?

**Comment by maleadt** [comment]
Found https://github.com/actions-marketplace-validations/EnricoMi_download-buildkite-artifact-action?tab=readme-ov-file#configuration; I've added a token

**Comment by christiangnrd** [comment]
Do we want to save the `minimum`, the `median`, or the 'mean`?

**Review by tgymnich — APPROVED** [review]

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -95,5 +110,44 @@ steps:
...
+
+            println("+++ :julia: Benchmarking")
+            include("perf/runbenchmarks.jl")'
+        artifact_paths:
+          - "benchmarkresults.json"
+        agents:
+          queue: "juliaecosystem"
+          os: "macos"
+          arch: "aarch64"
+          macos_version: "15.0"
```
Comment (maleadt) [inline .buildkite/pipeline.yml:148]: Why do you need macOS 15 for the benchmarks?

**Review by maleadt — COMMENTED** [review]
I guess we'll only see this in action on subsequent PRs?

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (christiangnrd) [reply .buildkite/pipeline.yml:148]: I initially wanted to see the performance impact of logging. Seems like now it's only enabled when actually used so might not be worth it.

**Comment by christiangnrd** [comment]
> I guess we'll only see this in action on subsequent PRs?

Looking at the Lux PR I based this off of it seems like it.

**Pull request merged**